File: mobile/lib/data/repositories/server_discovery_repository.dart
import '../../domain/models/connection_target.dart';
import '../../domain/models/discovered_server.dart';
import '../services/discovery.dart';

/// Single source of truth for discovered servers.
///
/// Wraps the stateless [ServerDiscovery] service, caches the last sweep, and
/// exposes domain models to ViewModels.
class ServerDiscoveryRepository {
  ServerDiscoveryRepository({required this._discovery});

  final ServerDiscovery _discovery;
  List<DiscoveredServer> _cached = const [];

  /// Last discovery results as an immutable snapshot.
  List<DiscoveredServer> get servers => List.unmodifiable(_cached);

  /// Runs an mDNS sweep and caches the results.
  Future<List<DiscoveredServer>> refresh() async {
    final servers = await _discovery.discover();
    _cached = List.unmodifiable(servers);
    return servers;
  }

  /// Builds a manual target from user-entered host and optional port.
  static ConnectionTarget manualTarget({required String host, int? port}) =>
      ServerDiscovery.manualTarget(host: host, port: port);
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/data/repositories/ServerDiscoveryRepository.kt
package dev.fazrigading.wheeldeck.data.repositories

import dev.fazrigading.wheeldeck.data.services.ServerDiscovery
import dev.fazrigading.wheeldeck.domain.models.ConnectionTarget
import dev.fazrigading.wheeldeck.domain.models.DiscoveredServer

/// Single source of truth for discovered servers.
///
/// Wraps the stateless [ServerDiscovery] service, caches the last sweep, and
/// exposes domain models to ViewModels.
class ServerDiscoveryRepository(private val discovery: ServerDiscovery) {
    private var cached: List<DiscoveredServer> = emptyList()

    /// Last discovery results as a snapshot.
    val servers: List<DiscoveredServer>
        get() = cached.toList()

    /// Runs an mDNS sweep and caches the results.
    suspend fun refresh(): List<DiscoveredServer> = discovery.discover().also { cached = it }

    companion object {
        /// Builds a manual target from user-entered host and optional port.
        fun manualTarget(host: String, port: Int? = null): ConnectionTarget =
            ServerDiscovery.manualTarget(host = host, port = port)
    }
}
